fs/ftpfs.py

```diff
@@ -331,22 +345,28 @@ def ftp_url(self):
     @property
     def ftp(self):
         if self._ftp is None:
-            self._ftp = self._open_ftp()
+            self._ftp = self._open_ftp(self.encoding)
             self._welcome = self._ftp.getwelcome()
         return self._ftp
 
     @property
+    def encoding(self):
```

Comment:
Would it not be possible to detect and store the encoding in the `ftp` property? Just wondering the lazy detection of encoding would result in excessive requests.

Reply:
I think so, but in both cases it only makes one additional call compared to the `master` implementation I guess. I could move it into the `ftp` property to encapsulate it.

Reply:
I think it might simplify things if the encoding detection logic is moved in to the `ftp`property. If the encoding defaulted to `latin-1` then you wouldn't need to pass it to `open_ftp`.
 
I'm also struggling a bit to understand why you need to open a new connection to detect the encoding. I think you should be able to call "FEAT" on first connection, and cache the encoding on the FTPFS. No need to make a new connection object.

Reply:
I agree with the first point, I'll change that.

For the second point: you have to set the encoding prior to connecting to the server. If you connect to the server with the default (latin-1), then change to utf-8 after seing the `UTF8` feature, the server will still think you're sending latin-1 encoded strings. So you to reset the server and use the proper encoding, you have to reconnect to the server with the new encoding.

Reply:
I think I've figured out something here. What was confusing me is that, I don't believe the encoding is ever sent to the ftp server. So the server is always going to send the same data regardless of what `FTP.encoding` is set to.

I've looked at the [source](https://github.com/python/cpython/blob/master/Lib/ftplib.py) and what appears to be happening is that its `ftplib` itself is using the encoding when it calls `self.sock.makefile` in the `connect` method. That creates a file-like object with the encoding when you connect.

So that would explain why you can't change the encoding after you have connected. But only due to a quite artificial limitation in ftplib. Otherwise you could just do a FEAT, and then set the encoding in the same connection.

It's really unfortunate.

I can think of a fudge, but its not very nice.

If the encoding is 'utf-8', you could re-*encode* the paths as 'latin-1' (effectively getting back the same bytes you received), then *decode* them as 'utf-8'. That should give you properly decode paths.

Another possibility is to do this if you detect 'utf-8' encoding:

```
ftp.file = ftp.sock.makefile('r', encoding='utf-8')
```

Both are hacks, so I'm not sure they are worth it. Otherwise we may be stuck with the single additional request to get the encoding...